File: 03-sliding_window/00121-Best_Time_to_Buy_&_Sell_Stock/00121-best_time_to_buy_&_sell_stock.cpp
// 121. Best Time to Buy and Sell Stock
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

#include <iostream>
#include <vector>

using namespace std;


class Solution {
public:
    // SOLUTION
    int maxProfit(vector<int>& prices) {
        int minValue = prices[0];
        int result = 0;

        for (int i=1; i<prices.size(); i++) {
            minValue = min(minValue, prices[i]);
            result = max(result, prices[i] - minValue);
        }

        return result;
    }
};


int main() {
    Solution o;

    // INPUT
    vector<int> prices = {7,1,5,3,6,4};

    // OUTPUT
    auto result = o.maxProfit(prices);
    cout<<result<<endl;

    return 0;
}

File: 03-sliding_window/00121-Best_Time_to_Buy_&_Sell_Stock/00121-best_time_to_buy_&_sell_stock.py
# 121. Best Time to Buy and Sell Stock
# https://leetcode.com/problems/best-time-to-buy-and-sell-stock/


class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        minValue = prices[0]
        result = 0

        for i in range(len(prices)):
            minValue = min(minValue, prices[i])
            result = max(result, prices[i]-minValue)

        return result



if __name__ == "__main__":
    o = Solution()

    # INPUT
    prices: list[int] = [7,1,5,3,6,4]

    # OUTPUT
    result = o.maxProfit(prices)
    print(result)
